jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/completion/DotNetHandlerCompletion.kt

```diff
@@ -30,31 +31,38 @@ class DotNetHandlerCompletion : HandlerCompletion {
         CamelHumpMatcher(prefix)
 
     override fun getLookupElements(project: Project): Collection<LookupElement> {
-        val model = project.solution.lambdaPsiModel
-
-        val lifetime = project.defineNestedLifetime()
-        val requestId = handlerRequestId.getAndIncrement()
-
-        var handlersValue: List<HandlerCompletionItem>? = null
-        model.determineHandlersResponse.adviseOn(lifetime, SynchronousScheduler) { response ->
-            if (response.requestId == requestId) {
-                handlersValue = response.value
+        val completionItems = getHandlersFromBackend(project)
+        return completionItems.map { completionItem ->
+            LookupElementBuilder.create(completionItem.handler).let {
+                if (completionItem.iconId != null) it.withIcon(FrontendIconHost.getInstance(project).toIdeaIcon(completionItem.iconId))
+                else it
             }
         }
+    }
 
-        model.determineHandlersRequest.fire(DetermineHandlersRequest(requestId))
+    @TestOnly
+    fun getHandlersFromBackend(project: Project): List<HandlerCompletionItem> {
+        val model = project.solution.lambdaPsiModel
+        val lifetime = project.defineNestedLifetime()
 
-        spinUntil(handlerRequestTimeoutMs) { handlersValue != null }
+        return try {
+            val requestId = handlerRequestId.getAndIncrement()
 
-        val completionItems = handlersValue
-            ?: throw IllegalStateException("Timeout after $handlerRequestTimeoutMs ms waiting for available handlers calculation")
+            var handlersValue: List<HandlerCompletionItem>? = null
 
-        lifetime.terminate()
-        return completionItems.map { completionItem ->
-            LookupElementBuilder.create(completionItem.handler).let {
-                if (completionItem.iconId != null) it.withIcon(FrontendIconHost.getInstance(project).toIdeaIcon(completionItem.iconId))
-                else it
+            model.determineHandlersResponse.adviseOn(lifetime, SynchronousScheduler) { response ->
+                if (response.requestId == requestId) {
+                    handlersValue = response.value
+                }
             }
+
+            model.determineHandlersRequest.fire(DetermineHandlersRequest(requestId))
+
+            spinUntil(handlerRequestTimeoutMs) { handlersValue != null }
+
+            handlersValue ?: throw IllegalStateException("Timeout after $handlerRequestTimeoutMs ms waiting for available handlers calculation")
+        } finally {
+            lifetime.terminate()
```

Comment:
@sdubov This would not terminate if the `handlersValue` was null before. Is this correct?

Reply:
Your change is valid. We need to manage lifetime ourselves using this API and make sure lifetime is terminated after timeout.
Could you please update `DotNetLambdaHandlerResolver::isMethodExists` function call with the same change?

Reply:
Thanks for confirming, will put out a PR to fix that one too

Reply:
#1342 fixes it